### PR TITLE
Improve ConnectionFailed exception error message

### LINKED ISSUE #1959: Send additional information when request evaluation is unable to connect to a service.
This is a follow-up issue for the PR #1939. When a request is sent for evaluation and linkerd is unable to connect to a downstream service. We get a message like:
```bash
$ curl localhost:4140 -H "Host: foo" -H "l5d-req-evaluate: 1"
connection timed out: /1.2.3.4:8888 at remote address: /1.2.3.4:8888. Remote Info: Not Available⏎
```
It would be better if we received information similar to when a request succeeds and is evaluated. With the additional message that says that the request was not able to reach the downstream service.

### PR DESCRIPTION
**Pull request by adleong**
The ConnectionFailed error message does not provide very much information, making it sometimes hard to tell why Linkerd is attempting to connect to that address in the first place.

We update the error message with contextual routing information, including much of the same information that is displayed in a diagnostic trace.

Old error message:
```
$ curl -v localhost:4140 -H "Host: default"
* Rebuilt URL to: localhost:4140/
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 4140 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 4140 (#0)
> GET / HTTP/1.1
> Host: default
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
< Via: 1.1 linkerd
< l5d-success-class: 0.0
< l5d-err: Connection+refused%3A+%2F127.0.0.1%3A8888+at+remote+address%3A+%2F127.0.0.1%3A8888.+Remote+Info%3A+Not+Available
< l5d-retryable: true
< Content-Length: 98
< Content-Type: text/plain
<
* Connection #0 to host localhost left intact
Connection refused: /127.0.0.1:8888 at remote address: /127.0.0.1:8888. Remote Info: Not Available
```

New error message:
```
$ curl -v localhost:4140 -H "Host: default"
* Rebuilt URL to: localhost:4140/
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 4140 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 4140 (#0)
> GET / HTTP/1.1
> Host: default
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
< Via: 1.1 linkerd
< l5d-success-class: 0.0
< l5d-err: Unable+to+establish+connection+to+127.0.0.1%3A8888.%0A%0Aservice+name%3A+%2Fsvc%2Fdefault%0Aclient+name%3A+%2F%25%2Fio.l5d.localhost%2F%23%2Fio.l5d.fs%2Fdefault%0Aaddresses%3A+%5B127.0.0.1%3A8888%5D%0Aselected+address%3A+127.0.0.1%3A8888%0Adtab+resolution%3A%0A++++%2Fsvc%2Fdefault%0A++++%2F%23%2Fio.l5d.fs%2Fdefault+%28%2Fsvc%3D%3E%2F%23%2Fio.l5d.fs%29%0A++++%2F%25%2Fio.l5d.localhost%2F%23%2Fio.l5d.fs%2Fdefault+%28SubnetLocalTransformer%29%0A
< Content-Length: 341
< Content-Type: text/plain
<
Unable to establish connection to 127.0.0.1:8888.

service name: /svc/default
client name: /%/io.l5d.localhost/#/io.l5d.fs/default
addresses: [127.0.0.1:8888]
selected address: 127.0.0.1:8888
dtab resolution:
    /svc/default
    /#/io.l5d.fs/default (/svc=>/#/io.l5d.fs)
    /%/io.l5d.localhost/#/io.l5d.fs/default (SubnetLocalTransformer)
* Connection #0 to host localhost left intact
```

Based on #1988 
Fixes #1959

### CODE REVIEW

File: router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import io.buoyant.namer.{DelegateTree, Delegator}
+import java.net.SocketAddress
+import scala.util.control.NoStackTrace
+
+/**
+ * An exception indicating that a connection could not be established to a remote peer.  This
+ * exception also includes additional information about the routing context for the purposes of
+ * displaying a more useful error message.  Since this exception is typically thrown in a router's
+ * client stack, the service name is not available and the `withPath` method can be used to
+ * argument this exception with the service name.
```
Comment (dadjeibaah) [inline router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala]: Is this suppose to be "argument" or "augment"?

File: router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+object RichConnectionFailedExceptionWithPath {
+
+  def apply(
+    path: Dst.Path,
+    client: Option[Name.Bound],
+    remote: SocketAddress,
+    namer: NameInterpreter
+  ): Future[Nothing] = {
+
+    val addresses = client match {
```
Comment (dadjeibaah) [inline router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala]: This code is very similar to [formatCtx](https://github.com/linkerd/linkerd/pull/1988/files#diff-7f2b008c320604fb7424f5702567777cR16). Might be a good idea to DRY up some of the code like `val resolution` or `val addresses`. I don't see this as a PR blocker.

Comment (adleong) [reply router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala]: possibly, although this code is short enough that I'm not sure how much there is to be gained by factoring it out.

File: router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  with FailureFlags[RichConnectionFailedExceptionWithPath] {
+
+  private[this] val addressList = addresses.collect {
+    case Address.Inet(isa, _) => isa.toString.stripPrefix("/")
+  }.mkString("[", ", ", "]")
+
+  private[this] val resolutionList = resolution.mkString("\n")
+
+  override def exceptionMessage(): String =
+    s"""Unable to establish connection to $remote.
```
Comment (olix0r) [inline router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala]: should this be val'd? or lazy val'd?

**Pull request merged**